.github/actions/upload/action.yml

```diff
@@ -1,31 +1,41 @@
 name: upload
-description: "Upload test results"
+description: "Upload test results, reports, and heap dump."
 inputs:
   name-suffix:
     default: ${{ github.job }}
+  upload-heap-dump:
+    default: false
   test-report-retention-days:
     default: 5
+  heap-dump-retention-days:
+    default: 14
 
 runs:
   using: composite
   steps:
     - name: Upload test results
       uses: actions/upload-artifact@v3
-      # Upload all test reports only on failure, because the artifacts are large
-      if: failure()
+      if: inputs.upload-results
       with:
         name: result ${{ inputs.name-suffix }}
+        if-no-files-found: 'ignore'
         path: |
           **/target/surefire-reports
           **/target/checkstyle-*
     - name: Upload test report
       uses: actions/upload-artifact@v3
-      # Always upload the test report for the annotate.yml workflow,
-      # but only the single XML file to keep the artifact small
-      if: always()
```

Comment:
if the previous upload fails (for any reason) - this upload won't run

Reply:
If I understand what you're trying to do (upload test results and heap dump only on failure, upload annotate results always) then you can reorder the steps like:

- upload test results (`if: failure()`)
- upload heap dump (`if: failure()`)
- upload annotate results (`if: always()`)

Reply:
The `failure()` and `always()` functions inside a composite action don't include state from outside. That's why I removed them, but good point on upload failures. I'll just add `always()` in every step.

Reply:
> The failure() and always() functions inside a composite action don't include state from outside.

Today I learnt - thanks.